langchain/chains/llm.py

```diff
@@ -115,20 +129,58 @@ async def aprep_prompts(
 
     def apply(self, input_list: List[Dict[str, Any]]) -> List[Dict[str, str]]:
         """Utilize the LLM generate method for speed gains."""
-        response = self.generate(input_list)
-        return self.create_outputs(response)
+        response, prompts = self.generate(input_list)
+        return self.create_outputs(response, prompts)
 
     async def aapply(self, input_list: List[Dict[str, Any]]) -> List[Dict[str, str]]:
         """Utilize the LLM generate method for speed gains."""
-        response = await self.agenerate(input_list)
-        return self.create_outputs(response)
+        response, prompts = await self.agenerate(input_list)
+        return self.create_outputs(response, prompts)
+
+    def _get_final_output(self, completion: str, prompt_value: PromptValue) -> Any:
+        """Validate raw completion (guardrails) + extract structured data from it (parser).
 
-    def create_outputs(self, response: LLMResult) -> List[Dict[str, str]]:
+        We may want to apply guardrails not just to raw string completions, but also to structured parsed completions.
+        For this 1st attempt, we'll keep this simple.
+
+        TODO: actually, reasonable to do the parsing first so that guardrail.evaluate gets the reified data structure.
+        """
+        for guardrail in self.guardrails:
+            evaluation, ok = guardrail.evaluate(prompt_value, completion)
```

Comment:
why is this better than `completion = guardrail.evaluate(prompt_value, completion)` and just letting the guardrail raise its own error?

Reply:
- i guess this is ", ok" from go, which is a nice idiom
- the EAFP/pythonic way is probably to raise an GuardrailException
- the flow control anywhere guardrails are used then will look like

```
try:
  evaluation = guardrail.evaluation(prompt, completion) # keep GuardrailEvaluation, easier to extend later
  if evaluation.revised_completion:
    completion = evaluation.revised_completion
except GuardrailException as e:
  raise e or something
```
 
- seems like the caller has to know a lot about this guardrail concept. for example, a completion may not initially satisfy the guardrail till after fixing/retrying the completion.. we have to check `evaluate`'s return value to know.

Wdyt is best?